libsolidity/ast/Types.h

```diff
@@ -973,6 +973,7 @@ class StructType: public ReferenceType
 	// Caches for interfaceType(bool)
 	mutable std::optional<TypeResult> m_interfaceType;
 	mutable std::optional<TypeResult> m_interfaceType_library;
+	mutable bool recursionGuard = false;
```

Comment:
Adding a mutable type just for checking whether the struct contains a mapping sounds a little weird.

Maybe we can disentangle this whole thing a little as follows:
 - remove `canLiveOutsideStorage` and replace it by a function `containsMapping` that is maybe only defined on ReferenceTypes if that simplifies it. I did a quick check on the code and I think that this function is mainly used on variables. Variables need a type that is written down and `canLiveOutsideStorage` for types that can be written down reduces to "contains mapping" - please check that again.
 - deduce the property of containing a mapping for structs at an earlier stage, maybe in a similar way as we determine whether a struct is recursive.

Reply:
@chriseth About removing `canLiveOutsideStorage`: there is an issue that talks about it previously, https://github.com/ethereum/solidity/issues/4804.

I tried to look into it. One problem I see is how extensively it is defined, for example even `RationalNumberType::canLiveOutsideStorage()` (which just returns false.) I'm not sure if removing these will cause any issues.

The second point about deducing this during annotation phase is doable, I think. We could add a new member in `StructAnnotation` called `containsNestedMapping`. What do you think?

Reply:
If you remind me, I can check next week whether `canLiveOutsideStorage` can be fully removed. In the meantime, you can implement the annotation member if you like.